lib/hal/video.c

```diff
@@ -331,12 +331,28 @@ void XVideoInit(DWORD dwMode, int width, int height, int bpp)
 
 	XVideoSetVideoEnable(FALSE);
 
+	if (framebufferMemory != NULL) {
+		MmFreeContiguousMemory(framebufferMemory);
+	}
+	framebufferMemory = MmAllocateContiguousMemoryEx(screenSize,
+	                                                 0x00000000, 0x7FFFFFFF,
+	                                                 0x1000,
+	                                                 PAGE_READWRITE |
+	                                                 PAGE_WRITECOMBINE);
+	assert(framebufferMemory != NULL);
+	memset(framebufferMemory, 0x00, screenSize);
+	asm __volatile__("sfence");
+
 	do
 	{
 		Step = AvSetDisplayMode((PVOID)VIDEO_BASE, Step, 
-			dwMode, dwFormat, pitch, VIDEO_FRAMEBUFFER);
+			dwMode, dwFormat, pitch, (unsigned int)framebufferMemory & 0x7FFFFFFF);
 	} while(Step);
 
+	/* Store the framebuffer that we set; normally this is done in XVideoSetFB.
+	   However, the first one is set using the kernel, so we do it here, too. */
+	_fb = framebufferMemory;
```

Comment:
This is a bit confusing as it could mean:
- we own two fb, one made by ourselves and one given to us by the xbox kernel
- we own one fb and the kernel does, but we do the same operation to our fb which thee kernel did to its fb
- we replace a fb given by the kernel with our own
So the comment expects abstract understanding about what's going on in the greater scope. This can be fine but I think giving it a second thought would be a good idea.
(It might be my incompetence here 🤷‍♀️)

Reply:
Neither of that. I agree that the wording is very poor.

What I meant is: We always store the *displayed* framebuffer *address* in `_fb`. Normally this is done in `XVideoSetFB` which sets the address to be displayed by the GPU and then keeps a copy of the set address in `_fb`.
However, when also changing the mode, we don't use `XVideoSetFB`. Instead, we are using the kernel function `AvSetDisplayMode` which internally also sets the address that will be displayed by the GPU; so we have to modify `_fb` after that, too.

I'll have to find a more concise way of phrasing this though.

I'll fix this when I get back to this branch; so keep this unresolved for now.